src/awsGuard.ts

```diff
@@ -0,0 +1,184 @@
+// Copyright 2016-2019, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import {
+    EnforcementLevel,
+    Policies,
+    PolicyPack,
+    ResourceValidationPolicy,
+    StackValidationPolicy,
+} from "@pulumi/policy";
+
+import { defaultEnforcementLevel, isEnforcementLevel } from "./enforcementLevel";
+
+const defaultPolicyPackName = "pulumi-awsguard";
+
+// Internal map that holds all the registered policy factories.
+const factoryMap: Record<string, PolicyFactory<any>> = {};
+
+/**
+ * A policy pack of rules to enforce AWS best practices for security, reliability, cost, and more!
+ *
+ * Create an instance of AwsGuard without any arguments to enable all policies with their default configuration:
+ *
+ * ```typescript
+ * const awsGuard = new AwsGuard();
+ * ```
+ *
+ * The above is equivalent to:
+ *
+ * ```typescript
+ * const awsGuard = new AwsGuard({ all: "mandatory" });
+ * ```
+ *
+ * To make all policies advisory rather than mandatory:
+ *
+ * ```typescript
+ * const awsGuard = new AwsGuard({ all: "advisory" });
+ * ```
+ *
+ * To make all policies advisory, but changing a couple to be mandatory:
+ *
+ * ```typescript
+ * const awsGuard = new AwsGuard({
+ *     all: "advisory",
+ *     ec2InstanceNoPublicIP: "mandatory",
+ *     elbAccessLoggingEnabled: "mandatory",
+ * });
+ * ```
+ *
+ * To disable a particular policy:
+ *
+ * ```typescript
+ * const awsGuard = new AwsGuard({
+ *     ec2InstanceNoPublicIP: "disabled",
+ * });
+ * ```
+ *
+ * To disable all policies except ones explicitly enabled:
+ *
+ * ```typescript
+ * const awsGuard = new AwsGuard({
+ *     all: "disabled",
+ *     ec2InstanceNoPublicIP: "mandatory",
+ *     elbAccessLoggingEnabled: "mandatory",
+ * });
+ * ```
+ *
+ * To specify configuration for policies that have it:
+ *
+ * ```typescript
+ * const awsGuard = new AwsGuard({
+ *     ec2VolumeInUseCheck: { enforcementLevel: "mandatory", checkDeletion: false },
+ *     encryptedVolumes: { enforcementLevel: "mandatory", kmsId: "id" },
+ * });
+ * ```
+ */
+export class AwsGuard extends PolicyPack {
+    constructor(args?: AwsGuardArgs);
+    constructor(name: string, args?: AwsGuardArgs);
+    constructor(nameOrArgs?: string | AwsGuardArgs, args?: AwsGuardArgs) {
```

Comment:
This may look odd, if you're unfamiliar with overloads in TypeScript. The first two signatures are the two available overloads, and the third is the implementation (the latter not being shown in intellisense).

This allows the following possible ways of constructing:

```typescript
new AwsGuard();
new AwsGuard({ all: "advisory"});
new AwsGuard("my-name", { all: "advisory" });
```